### ISSUE TITLE
Allow passing misc data to http client options

### ISSUE THREAD

**Issue by mattlewis92** [angular/angular#18155]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[x] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

Currently when using a http interceptor you cannot pass meta data about the request to the interceptor

## Expected behavior

To be able to pass random data to the http interceptor. Currently the options object is so strongly typed that you cannot pass extra data to it. The proposal would be to add a `meta` (name needs bikeshedding) property with type `any` to the http request options that can be used for this purpose.

## Minimal reproduction of the problem with instructions
N/A

## What is the motivation / use case for changing the behavior?

```
// tell a caching interceptor to only cache if this is true
this.http.get('/path', {meta: {cache: true}})

// tell an interceptor not to add the `Authorization` for this request only
this.http.get('/path', {meta: {skipAuthorization: true}})
```

## Environment

<pre><code>
Angular version: 4.3.0

Browser:
- [x] Chrome (desktop) version 59
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX

Others:

</code></pre>

**Comment by jnizet** [comment]
This would also be very useful for generic error interceptors:

    // tell an interceptor that 404 error is expected and shouldn't be displayed
    this.http.get('/path', {meta: {skipErrorHandling: resp => resp.status === 404}})

I see plenty of use-cases for this, and it was something possible in AngularJS using the options object.

**Comment by alxhub** [comment]
This has been on my mind as well. I think it's a good idea and something we should implement.

**Comment by mattlewis92** [comment]
@alxhub awesome! What do you think of the proposed API, can you think of anything better? I'd love to take a go at a PR for this 😀

**Comment by alxhub** [comment]
@mattlewis92 I like `meta` (or possibly `metadata`) as the key. In keeping with the theme for `HttpRequest` and `HttpResponse` it should also be immutable a la `HttpHeaders` / `HttpParams`.

I have an open task to merge the implementation of the two by extracting out a generic immutable map type. Once this is done, `HttpMetadata` could easily be built on top of that type.

**Comment by mattlewis92** [comment]
Ah ok I'll leave that to you to implement then 😄

**Comment by Agrumas** [comment]
Is there any neat workaround for this? I am thinking about extending `HttpRequest`, however, overloaded constructors make implementation complicated.

**Comment by ChristianUlbrich** [comment]
@Agrumas You could abuse HttpHeaders. Set them on your request with a nice name e.g. `X-META-DATA` and then strip them in the interceptor from the actual request before it gets send and re-add them on response. That way I plan to enable disabling _individual_ interceptors on a per-request basis, e.g. for toggling a global LoadingNotification. But of course this feature needs to be added - because it has so many use cases.

**Comment by alxhub** [comment]
I have an in-progress branch for this, but it's unlikely to land before 5.0. Probably coming in 5.1.

**Comment by jbojcic1** [comment]
@alxhub what's the status of this? I am really looking forward to it.

**Comment by sinedied** [comment]
Really disappointed that with the release of Angular 5, the team deprecated the `Http` service without providing any way to perform with `HttpClient` what was possible with `Http`, including custom parameters.

We currently extends the `Http` service and added with 2 extra parameters to `RequestOptionsArgs`, for common use cases already cited here such as caching requests and error handling (see https://github.com/ngx-rocket/starter-kit/blob/master/src/app/core/http/http.service.ts for details).

We would like to move to `HttpClient`, but we cannot use interceptors because they do not support custom arguments, and we cannot extend `HttpClient` either because the `request()` method options do not use a type we can extend, see https://github.com/angular/angular/blob/master/packages/common/http/src/client.ts#L76-L83

So basically we are stuck with a deprecated API, and no way to move forward, am I missing something here @alxhub ?

**Comment by karptonite** [comment]
@sinedied I've been using HttpClient and adding metadata by abusing HTTP Headers, as suggested here.

In practice, it looks like this:
```ts
    if (req.headers.has('Skip-Prefix')) {
      const headers = req.headers.delete('Skip-Prefix');
      const directRequest = req.clone({ headers });
      return next.handle(directRequest);
    }
```

**Comment by sinedied** [comment]
@karptonite I saw the comment on this, but this is really a ugly hack and I would like to avoid those while promoting best practices in our generator 😉

**Comment by karptonite** [comment]
@sinedied It isn't pretty, but I've seen far uglier hacks. You can extract out the addition of the headers into a utility function so that when the metadata is added for real, you only have to fix it in a couple places. Anyway, you aren't stuck with a deprecated (but not yet removed) API--you have at least one option.

**Comment by ChristianUlbrich** [comment]
@sinedied For achieving this, you won't need to pass arbitrary params to the HttpClient (and thus to all interceptors); but you could wrap this behavior in a sugar method that clones your HttpService to a variant that is _always_ caching with a fluent API like:

```
   this.httpService
        .enableCache()
        .get('http://google.com')
``` 
That way the implementation is hidden away and you should be prepared when HttpClient is going to be _deprecated_ ...

Passing certain magic options along the request options looks like the bigger hack to me. _Or:_
>  It was fun while it lasted.

**Comment by wKoza** [comment]
Hi @alxhub , are you always ready to implement this feature in Angular 5.1 ?

**Comment by jcimoch** [comment]
Is it going to be merged?

**Comment by dawidgarus** [comment]
Proposed solution with `metadata` has a big drawback - no type checking. Also, one interceptor can mess with other's metadata and naming conflicts are possible.
What if we pass array of objects as metadata when we create HttpRequest (ex.: `[new MyInterceptorMetadata(params)]`)? Interceptor could get it's own metadata with `request.metadata.find(m => m instanceof MyInterceptorMetadata)` or even `request.metadata.get/has(MyInterceptorMetadata)`.
And when cloning, it should not be possible to change metadata, except for adding new objects perhaps.

**Comment by sinedied** [comment]
@dawidgarus The type checking issue could be easily resolved, if an empty type interface `RequestMetadat` is used instead of any. Then projects just have to augment this interface to add their own properties and still respect type checking.

As for immutability, the metadata object could be made immutable by using `Object.freeze()` or by passing a copy of the object to each interceptor.

**Comment by dawidgarus** [comment]
@sinedied Empty interface doesn't provide any type checking. You could make a typo: `{metadata: { 
myIntrceptor: params }}` and You wouldn't know about it. `Object.freeze` is useless in this case - object cannot be mutated, but can be replaced when request is cloned. And there still is problem with naming conflicts. Two libraries could have interceptors which expects parameter with the same name. Maybe the odds are slim, but if it happens, there's no workaround. That's the reason why we use `InjectionToken` and not strings for DI.

**Comment by wKoza** [comment]
Concerning the problem of type checking, I have proposed, here #20579,  this:  
`metadata?: {[key: string]: any}`. 
So, you can have  keys that reflects each interceptor. 
For the clone aspect, it's the same principe that `HttpHeaders`

**Comment by Yura13** [comment]
When will this feature plan to release?

**Comment by sn0opr** [comment]
I'm starting a new project and I really like this feature to be released ASAP!

**Comment by dawidgarus** [comment]
It's easy to extend `HttpClient` and add custom methods for dynamic interceptors:
https://gist.github.com/dawidgarus/6fd4ae338334ba65fd5b37e1a74bb13e
Usage example:
```ts
httpService.noCache().get('http://example.com')
```
Ofc it's possible to pass own parameters to custom methods and interceptors

**Comment by mburger81** [comment]
@alxhub are there a roadmap or any plan on this?

**Comment by jcimoch** [comment]
@dawidgarus The point is not to do this. It would be nice to be able to send metadata without extending HttpClient. There are plenty of reasons why you would want to inject raw httpClient and not your custom implementation. For example complex project which is combination of different modules across different teams which are developing modules in isolation of consuming app would force them to apply every minor change to keep all extended httpClient classes in sync, provide one really big and ugly or make them every time yo want to inject it in some module with some customization. Using native implementation can solve all of this problems.

**Comment by dawidgarus** [comment]
@jcimoch It's still possible to achieve that. I ended up doing something like this in my project: https://gist.github.com/dawidgarus/85e93a3dcb9e35fc7a2038484f3512b5
All teams just have to agree to use `HTTP_SERVICE_INTERCEPTORS` instead of `HTTP_INTERCEPTORS`

**Comment by rtrompier** [comment]
Hi guys,
Any news about this feature ?

**Comment by RainingNight** [comment]
Hi guys,
Any news about this feature ?

**Comment by sinedied** [comment]
@dawidgarus Thanks for sharing your solutions. After giving it some thoughts, I'm going with the fluent API and dynamic interceptors options.

**Comment by sinedied** [comment]
Finally went around it, it took some serious work and mixup of proposed solutions here, as I needed to:
- Keep backward compatibility with `HttpClient` so the new service can be injected in place of it
- Dynamically add or remove (that was the hardest!) and configure interceptors per request
- Make it convenient for the users to reduce the boilerplate code

I came up with this: https://gist.github.com/sinedied/46f9b3cf5cbd04fa452d1446edc6006d

I'm not completely satisfied with this solution, but at least it allows to use standard interceptors and limit the breaking changes for the users (move to options in the request itself to add option calls on the service).
If you have better ideas on the approach I'm taking it...

And seriously, extending `HttpClient` methods is really painful because of all the overloads, I had to resolve to use `any` types to abuse TypeScript to allow me to override the `request()` method, which I find really awful 😞

**Comment by FDIM** [comment]
I'll join the party and agree that `meta` (or `metadata`) option is a needed feature to effectively use interceptors. As @alxhub suggests, empty `HttpMetadata` interface that would be extended locally in http-client.d.ts file would work for our use case.
To avoid clashing of options between interceptors, @dawidgarus maybe this should be left to the authors to have an optional dependency that would dictate which actual option it should use? This is a non issue for us as the interceptors is a 'core' feature and maintained by one team.

Aside of that, there is this example in http client docs about caching: https://angular.io/guide/http#caching - does anyone have any input on possible implementation of `isCachable` function ? I can only think of a url -> boolean map that would be configured somewhere else, but this is far from optimal.

**Comment by JWess** [comment]
I would love to have this ticket addressed! In the meantime, here is a pretty decent work-around that I came up with:

https://stackoverflow.com/a/49903921/1011648

**Comment by iwantwin** [comment]
+1 have used headers to work around this so many times, found this issue and it needs more love :)

**Comment by ndcunningham** [comment]
Is there any update on this. It would be really beneficial to know the status.

**Comment by FDIM** [comment]
I would really like to have this feature in and I think I have a solution that would fit everyone.

So how about we add following classes/interfaces to http client package and additional meta/metadata property when making requests:
```
export interface InterceptorOption<T> {
  Type: T
}
//////////////
export class InterceptorOptionToken<T> extends InjectionToken<T> implements InterceptorOption<T>  {
  Type: T;
} // doesn't have to extend InjectionToken
//////////////
export class InterceptorMetadata<T extends InterceptorOption<any>> {

  constructor(key?: T, value?: T['Type']) {
    if (key && value) {
      this.put(key, value);
    }
  }

  put<T extends InterceptorOption<any>>(key: T, value: T['Type']): InterceptorMetadata<T> {
    // todo: store details in a map, injector? weakmap?
    return this;
  }  
}
```
Usage in library:
```
// my custom interceptor with an option
export const CacheInterceptorOptions = new InterceptorOptionToken<{cache: boolean}>('cache');
@Injectable()
export class CacheInterceptor { ... }
```

Usage in services / components:
```
// compiles
this.http.get('/path', { meta: new InterceptorMetadata(CacheInterceptorOptions, { cache: true }) })
this.http.get('/path', { meta: new InterceptorMetadata(CacheInterceptorOptions, { cache: false }) })

// error
this.http.get('/path', { meta: new InterceptorMetadata(CacheInterceptorOptions, { 
  foo: 123 // Argument of type '{ foo: number; }' is not assignable to parameter of type '{ cache: boolean; }'.
})}) 

// error
const m = new InterceptorMetadata();
m.put(CacheInterceptorOptions, 123); // Argument of type '123' is not assignable to parameter of type '{ cache: boolean; }'.
// Note that 'put' can be chained
```

I am also interested in making a PR.

**Comment by rahul-winner** [comment]
Are we submitting this PR to Angular 6 ?

**Comment by FDIM** [comment]
Given how much attention there was from contributors I'd say no. I just hope that it will land in 7.

**Comment by lazarljubenovic** [comment]
@alxhub According to the comments here, it seems like you've been working on this. There's also [wKoza's PR](https://github.com/angular/angular/pull/20579) which seems stale and [another one](https://github.com/angular/angular/pull/25751) is a bit newer, however it seems like not all tests are passing.

Do you happen to have info on the progress? It's been almost a year since the last update.

**Comment by wKoza** [comment]
I can update this PR if necessary. I worked on another [PR](https://github.com/angular/angular/pull/26019) which allow having multiple instances of HttpClient.

**Comment by FDIM** [comment]
@lazarljubenovic The tests that are not passing are flaky ones and unrelated to the changes I did. Aside of that the solution in my PR is type safe :)

**Comment by dawidgarus** [comment]
@FDIM your solution is a little bit overengineered.
Something like this would be much simpler, less verbose and still type safe:
```typescript
export class CacheInterceptorOptions {
  constructor(public cache: boolean) {}
}

@Injectable()
export class CacheInterceptor { 
  intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
    const cacheOptions = req.getMeta(CacheInterceptorOptions);
    if (cacheOptions && cacheOptions.cache) { ... }
  }
}
```
and usage:
```typescript
this.http.get('/path', { meta: [new CacheInterceptorOptions(true)] })
```

**Comment by sebastian-zarzycki-es** [comment]
So, when can we get the functionality that is available since Angular 1.x, basically for years now, in modern Angular? It seems bizarre that adding a simple field, heck, make it untyped, for what I care, is causing so much discussion and trouble. Just add it, for pete's sake. If you think it's wrong solution, then what's stopping you from switching to new one ONCE it is ready? But why oh why we cannot have the simple version now, WHILE we are waiting?

**Comment by lazarljubenovic** [comment]
> adding a simple field, heck, make it untyped, for what I care,

Because we care too.

> But why oh why we cannot have the simple version now, WHILE we are waiting?

Because it's a recipe for a disaster.

**Comment by FDIM** [comment]
Solution from @dawidgarus is nice as well, how could we get some attention to this issue from someone that could make a decision and also has permission to merge? I am not picky for verbosity as long as it is type safe.

**Comment by jburtondev** [comment]
Hi @alxhub, do we have an update for this?

**Comment by dhilgarth** [comment]
Based on the suggestions in this issue I have created a simple solution that can be used until there is first class support for this in Angular: https://gist.github.com/dhilgarth/783ed994be6e3af0c236fc36f327c9cd
Maybe it helps someone. Feedback welcome.

**Comment by Agrumas** [comment]
Huh, #25751 attracted some attention, fingers crossed 2019 will be a good year and this feature lands

**Comment by bh3605** [comment]
I'm working on a workaround that extends off of the base HttpRequest object where additional properties are defined.

    export class HttpRequestExtended<T> extends HttpRequest<T> {
    public needAssessmentsToken: boolean;
    public service: any;
    public needAuthToken: boolean;

    public constructor(method: 'DELETE' | 'GET' | 'HEAD' | 'JSONP' | 'OPTIONS' | 'POST' | 'PUT' | 'PATCH',
                       service: any,
                       body: T | null,
                       init?: {
                           headers?: HttpHeaders;
                           reportProgress?: boolean;
                           params?: HttpParams;
                           responseType?: 'arraybuffer' | 'blob' | 'json' | 'text';
                           withCredentials?: boolean;
                       }
    ) {
        super(method, service.url, body, init);
        this.service = service;
        this.needAssessmentsToken = service.needAssessmentsToken;
        this.needAuthToken = service.needAuthToken;
    }`

You then have a "wrapper" where this HttpRequest subclass is typesafe

    export class HttpClientExtended {
    public constructor(private httpClient: HttpClient) {
    }
      /**
       * GET request
       * @param {string} endPoint backend endpoint to call
       * @param {IRequestOptions} options request options (headers, body, etc.)
       * @returns {Observable<T>}
       */
      public get<T>(arr: any, options?: RequestOptions): any {
        const request: HttpRequestExtended<any> = new HttpRequestExtended('GET', arr, null, options);
        return this.httpClient.request(request).pipe(
          filter((event: HttpEvent<any>) => event instanceof HttpResponse),
          map((response: HttpResponse<any>) => response.body));
      }

Now your service looks like

    constructor(private httpClient: HttpClientExtended) {
    }

    private getItems(): Observable<any> {
        return this.httpClient.get({ url: 'my-url', needAuthToken: true});
    }

In my actual implementation that object is actually a getter property on an object that setups the appropriate url and whatever custom data I want to pass to my interceptors

Now your intercept method in your Interceptor looks like:

    intercept(req: HttpRequestExtended<any>, next: HttpHandler): Observable<HttpEvent<any>> {
    if (req.needAuthToken) {

One thing to point out is I have to override HttpRequest's clone method so my custom properties remain any time an interceptor calls clone() to add or remove HttpHeaders

      clone(update: {
    headers?: HttpHeaders,
    reportProgress?: boolean,
    params?: HttpParams,
    responseType?: 'arraybuffer'|'blob'|'json'|'text',
    withCredentials?: boolean,
    body?: any|null,
    method?: string,
    url?: string,
    setHeaders?: {[name: string]: string | string[]},
    setParams?: {[param: string]: string};
     } = {}): HttpRequestExtended<any> {
      var newRequest = super.clone(update);
      return new HttpRequestExtended(newRequest.method as 'DELETE' | 'GET' | 'HEAD' | 'JSONP' | 'OPTIONS' | 'POST' | 'PUT' | 'PATCH', this.service, newRequest.body, Object.assign({}, newRequest));
  }

In my approach I'm choosing to pass in an HttpRequest object, but HttpClient handles HttpRequest objects differently from a url, body, options parameters. This is causing me to essentially rewrite HttpClient's request method response type switch case (which isn't shown above since I haven't written it). This left me wondering if a different approach would be to allow me to pass in an HttpRequest object with a property that determines whether I want to listen to the HttpEvent stream or the response body itself. HttpClient wouldn't have to recreate the HttpRequest object thus preserving any custom properties I have defined. I wouldn't need the HttpClientExtended class anymore since I'd be able to pass in any HttpRequest type I want.

**Comment by maxandriani** [comment]
Hi guys. Do we have some news about this issue?

**Comment by ricardo-mello** [comment]
Based on previous solutions, I did an HttpService that extends HttpClient and use it on my entire application: 

https://gist.github.com/ricardo-mello/849a205ef17c2754ba819908f6fd4efb

Hope it helps someone 😄

**Comment by tommck** [comment]
Really sucks that we're STILL left with the choice of hacking headers or wrapping the HttpClient

**Comment by Agrumas** [comment]
It's especially sad that good enough solution has been already proposed, implemented and PR created. Overall, I love Angular because it provides a great approach to build apps. It's just this one limitation that requires ugly workarounds for the trivial use-case. I am pretty sure that anything would be better than having nothing for nearly 3 years.
Anyway, I am extremely grateful to the Angular team for this framework.

**Comment by aksdevac** [comment]
This should be added in v11.

**Comment by sebastian-zarzycki-es** [comment]
This should be added in v2.

**Comment by CesarD** [comment]
I can't believe we're still missing this SO REALLY BASIC feature! This is something even AngularJs had!

**Comment by david0603** [comment]
bttt

**Comment by jcimoch** [comment]
By the way if someone think thinks it is only one of two related threads of this feature...
Back in 2017 I created a feature request about "ignorable http interceptor" but in fact allowning to pass metadata is just extension of this idea to more generic approach.

https://github.com/angular/angular/issues/20203#issuecomment-347703209

As you can see over two years and high demand of that feature is still not enough to convince some core contributors

**Comment by Lonli-Lokli** [comment]
@jcimoch there was a discussion some time ago, on one of ngconf this issue has been mentioned. As I understand, there was no agreement on design of this , eg if this metadata should be readonly or not.
@mhevery can you reference appropriate personal if he still googler?

**Comment by maxandriani** [comment]
@alxhub Is there something we can do to Help on this issue? I saw there is an PR that you was onvolved that looks great. Why you guys can not aprove it? I know there is a issue about patterns that should be argued. But we live on agile environment w/ incremental features. You can release a initial version and fix it as the problem happens.

**Comment by sebastian-zarzycki-es** [comment]
Operation Byelog eh? I'm sorry to be harsh, but if Angular team is focusing on pretty minor (nowadays cpu/internet performance considered) speed/size improvements and dances around engines and compilers for another couple years, rather that adding meaningful features, then something is way off balance there.

**Comment by jcimoch** [comment]
@sebastian-zarzycki-es I have exactly the same feeling over years of using angular now. It seems like there are couple of people that are fascinated over compilers, performance, alogrithms but not interested in making usefull framework solving real business problems. Maybe if angular compiler wasn't over engineered in the first place (because AOT in its initial format bascially has taken best advantages of dynamic javascript environment in favour of some perfomances gains which were at the best only close to react performance which is fully runtime based). Now with ivy we are bringing back basic stuff like using arrow function were I want and not were angular compiler allows... or dynamically create/load component fetched from resource out of the project, this wasn't possible in any reasonable way before. So we are solving issues that we created and were solved long time ago, even in angularjs. 
Something is way off balance there as you said.

**Comment by CesarD** [comment]
Brace yourselves to get some more comments minimized... 
Someone is more interested in minimizing requests and claims from actual users of the framework than in adding requested features.

**Comment by marcschroeder** [comment]
@alxhub is there anything we can do to support you on this issue? Who of the core team is assigned to this? I feel like there have been plenty of people offering there help over the past three years that this issue is open. This is a non-breaking change, right? What can we do to get this into the next minor release?

**Comment by aksdevac** [comment]
As per this comment (on related PR) from angular team member this is not coming anytime soon as it requires code refactoring in core:
https://github.com/angular/angular/pull/20579#issuecomment-736828812

But 'InterceptorSkipHeader' (although not the best) solution is working fine for me.
```ts
intercept(req: HttpRequest<unknown>, next: HttpHandler): Observable<HttpEvent<unknown>> {
  if (req.headers.has(InterceptorSkipHeader)) {
    // console.log('skiping intercepter');
    const headers = req.headers.delete(InterceptorSkipHeader);
    return next.handle(req.clone({ headers }));
  }
  // any other logic for adding common header
  // ...
}
```

**Comment by alxhub** [comment]
For anyone keeping up with this issue, discussions on PR #20579 have converged on a simple and straightforward design which we hope to be able to land in the next couple days :)

**Comment by sebastian-zarzycki-es** [comment]
4 years of waiting for a change of heart (for the context to be mutable) to eventually implement a simple typed map. Oh, the Angular team, you never cease to amaze me :)

(still, thanks for this)

**Comment by jcimoch** [comment]
![obraz](https://user-images.githubusercontent.com/5717461/106590144-17bbc180-654d-11eb-8d55-6ddfa19b7c75.png)

**Comment by alxhub** [comment]
Closing as fixed in #25751 (will land in 12!)

**Comment by CesarD** [comment]
> Closing as fixed in #25751 (will land in 12!)

Will we have to wait until Angular 12 to make use of this? That's around May, isn't it?
